### DISCUSSION TITLE
Accessing previous value in persist setItem

### DISCUSSION THREAD

**Discussion by TeemuKoivisto** [pmndrs/zustand#2876] [Q&A]
Hey, I've been a happy user for a zustand for some time now. Just now I've started to integrate my zustand stores with IndexedDB (through Dexie) and I noticed one rather annoying aspect of it.

Mainly, as I am using `import { persist } from 'zustand/middleware'` and `setItem: (name, newValue: StorageValue<State>) => {` to update my IndexedDB, there's no built-in way of accessing the previous value of the state.

This would be rather useful as I can't reliably check _what_ has changed so I have to run complicated `db.table.bulkPut` logic for every tiniest change that might not change anything in IDB. The performance implications of this might not be huge although with large in-memory datasets and lot of updates to the store, it could cause unnecessary CPU & memory use on mobile devices especially.

Could this value be provided as 3rd argument?

**Comment by dai-shi** [comment]
I wonder if creating a custom storage would work it around:
```js
const createCustomStorage = () => {
  let prevValue;
  return {
    getItem: (name) => {
      ...
    },
    setItem: (name, newValue) => {
      ...
      prevValue = newValue;
    },
    removeItem: (name) => {
      ...
    },
  };
};
```

**Reply by TeemuKoivisto** [reply]
Hey thanks for the quick reply! Yea that's actually what I ended up doing myself. Just assigned the IDB persisted list to a local variable which I check inside setItem and run the put logic based on it. But not very pretty.

**Reply by TeemuKoivisto** [reply]
Also why in `getItem` you can't just return a slice? Since I might want to leave some values as default. Just wondered this as I noticed that I had a faulty JSON.parse but getItem silently just ignored the error and failed the hydration.

**Reply by dai-shi** [reply]
> Also why in `getItem` you can't just return a slice?

I'm not sure if I follow it.

Are you talking about this getItem?
https://github.com/pmndrs/zustand/blob/5ed67ec1a9f8a5ee2f3aa19d5092d661cca4b126/src/middleware/persist.ts#L8

**Reply by TeemuKoivisto** [reply]
> > Also why in `getItem` you can't just return a slice?
> 
> I'm not sure if I follow it.
> 
> Are you talking about this getItem?
> 
> https://github.com/pmndrs/zustand/blob/5ed67ec1a9f8a5ee2f3aa19d5092d661cca4b126/src/middleware/persist.ts#L8

This one:

```ts
export interface PersistStorage<S> {
    getItem: (name: string) => StorageValue<S> | null | Promise<StorageValue<S> | null>;
    setItem: (name: string, value: StorageValue<S>) => unknown | Promise<unknown>;
    removeItem: (name: string) => unknown | Promise<unknown>;
}
```

Or has it changed? Seems there has been a major version released since I last updated zustand.

**Reply by dai-shi** [reply]
Ah, yes, this still exists:
https://github.com/pmndrs/zustand/blob/5ed67ec1a9f8a5ee2f3aa19d5092d661cca4b126/src/middleware/persist.ts#L18-L24

And, what do you mean by a slice?

**Reply by TeemuKoivisto** [reply]
I can't return a partial state eg `return { state: { items } }` where `state` should have other values as well. It shows `Types of property 'state' are incompatible.` even though I'd want to use the default values from the store. Again, accessing the current/previous state would be handy in the function parameters.

**Reply by dai-shi** [reply]
Have you considered using the `merge` option?

https://github.com/pmndrs/zustand/blob/2b7b09e2f9c8397c9267e734ddc7cd04332af6bf/src/middleware/persist.ts#L105-L109

**Reply by TeemuKoivisto** [reply]
It's not the same thing. It doesn't change the required return value of `getItem`. Besides, I'd only do: `merge: (v1, v2) => { ...v2, ...v1 }`

**Reply by dai-shi** [reply]
If you are talking about a TypeScript limitation of the required return value. I think you can explicitly specify it as `PersistedState`.

https://github.com/pmndrs/zustand/blob/2b7b09e2f9c8397c9267e734ddc7cd04332af6bf/src/middleware/persist.ts#L66

**Reply by TeemuKoivisto** [reply]
Of I course I can do that. I could set it `as any` for all what's worth.  Not going to test it whether it breaks, but I presume it doesn't. But using `as` is a code smell in TS as it breaks type guarantees in hidden ways. If I set it `as State` but it's not State, you can see how it can break things downstream. The fact that `PersistOptions` is passed the store as its generic type, would indicate that I do not have to mutate it in any shape. 

For this use case, making the S partial would solve it much more elegantly: `getItem: (name: string) => StorageValue<Partial<S>> | null | Promise<StorageValue<Partial<S>> | null>;` if zustand would want to allow hydrating partial states. Or passing down the oldState in the function, then you wouldn't have to make it Partial.

**Reply by dai-shi** [reply]
I'm not sure why you need `as any`. It's likely there are some misunderstandings of mine, but can you create a small reproduction with https://typescriptlang.org/play ?

**Reply by TeemuKoivisto** [reply]
Here's a sandbox https://codesandbox.io/p/sandbox/fwq5dr Inside `store.ts` you should get a TypeScript error. LocalStorage doesn't work inside an iframe so don't mind that

**Reply by dai-shi** [reply]
Here's the same one with TS playground: https://tsplay.dev/WYxAQm

Here's the modified one using `partialize`: https://tsplay.dev/WkZ4lm
(We don't need `merge` because the default merging is fine in this case.)

**Reply by TeemuKoivisto** [reply]
There is something seriously wrong with that type checking as this compiles [Playground Link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgCIAvAVwGcYMA7AEzIG4AoUSWROMNKDWB0ANHAAKAoXQDyYGMAgMaYgMoxoGAOZoAahgA2VHPiIly1OoyYB6EMCZMDaAO4Z0bduzQAPbvBgAT344dSwcAF5EdjhCCAgALjg6KGAGLQ5YgCN3JIYqECyBDlwOIJCAQWQFJRo4KIQY5LQYADF4gAoCeKSUtK0ASiSANwgHTLg01DQQNAYYACF3DqG4UfGm9BoWpagVkbGmEo5vP2h4ZFr4WjR1aEiUdHCAHjDsOAAyOCqa5QA+FYdJr8QTCGBA2KxDrbGBiHQwAb1P7RSGQ9AwKhQBgo1GQ7qJchMNAEDBUAwwMgiJq4uA5KBJAAMVJpsRh7QJXXiiIiyMaLMhMI6SHxeAGExpuGZLKm6Fm812SRWSJx-MF0Po2G5yKF1P5tNyyQ1aAAdHS4ABqOAARil-NwAzFutRkqdaLQMIVcCVPJVLLVWq9SDpjNFjrttshpV1LtRfNxDAwsySZDo90prtTUG0aCScZp8IAktgQIqE7MA3mWZdlPAzCACwARBb1OAAKVUMgAcsawO5th0DBBkIY7lmdMbC8WOmW0IiAD5z8hkeeLhnilnozHYyv8mw2OAwAAWwjg6AAjlRgFs4hAD7einAiQQ0mgmLSqPACwByN9ueZ3uAbldGk9wPQ8cCfUlyTWQxjEmGtMCYYDcSsbBc2QkD91ifFkzSIkfFfdM9VxHDrQjYjg3IOtGwWIjiLwcjcSjcNkJhIsZkVAB9GcxAYVx9CMWdlR3XFqzoXgzXwKI+JcATjGNVC0HXTC4AAAQAFRkBsZEmeAXAgMk3wfBhgGQHBgAIMCBBwNw6ggAwmDk8zsSPHB3C0Ao5ngAwwWNPyMLgUCQWkGB1IgGiOjpAYAsHYcDFHbMFJadiQA6bjEzQMR2y7BSYFSdILMCSL3AdZTI0Y2JZQgYY0BS0sMoDWKRw0McTSqmqUunBrGJjSMmmix0gA) But, I see there's a partialize method and second generic argument for the persisted state. 

Although if you have them defined as generics in the interface, you should use that instead of manually typing the function parameters. `satisfies` operator is basically made for this [Playground Link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgCIAvAVwGcYMA7AEzIG4AoUSWROMNKDWB0ANHAAKAoXQDyYGMAgMaYgMoxoGAOZoAahgA2VHPiIly1OoyYB6EMCZMDaAO4Z0bduzQAPbvBgAT344dSwcAF5EdjhCCAgALjg6KGAGLQ5YgCN3JIYqECyBDlwOIJCAQWQFJRo4KIQY5LQYADF4gAoCeKSUtK0ASiSANwgHTLg01DQQNAYYACF3DqG4UfGm9BoWpagVkbGmEo5vP2h4ZFr4WjR1aEiUdHCAHjDsOAAyOCqa5QA+FYdJr8QTCGBA2KxDrbGBiHQwAb1P7RSGQ9AwKhQBgo1GQ7qJchMNAEDBUAwwMgiJq4uA5KBJAAMVJpsRh7QJXXiiIiyMaLMhMI6SHxeAGExpuGZLKm6Fm812SRWSJx-MF0Po2G5yKF1P5tNyyQ1aAAdHS4ABqOAARil-NwAzFutRkqdaLQMIVcCVPJVLLVWq9SDpjNFjrttshpV1LtRfNxDAwsySZDo90prtTUG0aCScZp8IAktgQIqE7MA3mWZdlPAzCACwARBb1OAAKVUMgAcsawO5th0DBBkIY7lmdMbC8WOmW0IiAD5z8hkeeLhnilnozHYyv8mw2OAwAAWwjg6AAjlRgFs4hAD7einAiQQ0mgmLSqPACwByN9ueZ3uAbldGk9wPQ8cCfUlyTWQxjEmGtMCYYDcSsbBc2QkD91ifFkzSIkfFfdM9VxYNCGIesmwjcMMLwddnSogUWiLGZFQAfRnMQGFcfQjFnZUd1xas6F4M18CiLiXB44xjVQtA6NxUCAAEABUZAbGRJngFwIDJN8HwYYBkBwYACDAgQcDcOoIAMJgpOM7EjxwdwtAKOZ4AMMFjS8mjQJBaQYGUiBGwWDo6QGGjB2HAxR2zGSmKndjEzQMR2y7GSYFSdITMCUL3AdeTYhjDcZggYY0GYksvRnANIpHDQxxNWVSvKqdqoYorCuSLBhGfd0JCkME5F+GhXiNMRxEMgBrUbwjEL86S-P4-iacLHSAA)

But this doesn't seem optimal to me also. Well, for some reason `partialize` breaks with `satisfies` but now within `setItem` I can no longer see the complete state—only the persisted value. So if I have some local variable that I use for eg. caching the persisted values, I no longer can see it. As a side-note, doing this `PersistOptions<S, PersistedState = S>` and then `storage?: PersistStorage<PersistedState> | undefined;` `export interface PersistStorage<S> {` breaks my brain. Is `PersistStorage<S>` now PersistOptions S or PersistedState?

Anyway. I think I've had enough and just manually type the values and be done with it. I don't think either of us ends up any wiser.

**Reply by dai-shi** [reply]
Yeah, `{ foo: 1, bar: 'a' }` is assignable to `{ bar: string }`, so we have to do runtime type check. In this case, it's easier to avoid objects and use primitive values instead.

In summary, this is what I would do: https://tsplay.dev/NrZ0zN

Thanks for raising this up.

**Reply by TeemuKoivisto** [reply]
That looks better! Thanks for the help
